GAE/src/org/waterforpeople/mapping/app/web/DataProcessorRestServlet.java

```diff
@@ -1520,18 +1524,32 @@ private void assembleDatapointName(Long surveyId, Long surveyedLocaleId) {
         List<SurveyedLocale> locales = null;
         if (surveyedLocaleId != null) {
             // Single locale update
+            SurveyedLocale sl = slDao.getById(surveyedLocaleId);
+            if (sl == null) {
+                log.log(Level.WARNING, "SurveyedLocale not found: " + surveyedLocaleId);
+                return;
+            }
+            surveyGroupId = sl.getSurveyGroupId();
             locales = new ArrayList<>();
             locales.add(slDao.getById(surveyedLocaleId));
+        } else if (surveyGroupId != null) {
+            // Fetch all locales for this survey group
+            locales = slDao.listLocalesBySurveyGroupId(surveyGroupId);
         } else {
-            // get locales by createdSurveyId
-            locales = slDao.listLocalesByCreationSurvey(surveyId, Constants.ALL_RESULTS, null);
+            return;
+        }
+        
+        SurveyGroup sg = sgDao.getByKey(surveyGroupId);
+        if (sg == null) {
```

Comment:
I was thinking perhaps to make this cleaner you can already check for `surveyedlocaleid == null` , `surveyGroupId == null` and `sg == null` at the start of the method. Preconditions.